Core/Core/CourseSync/CourseSyncSelector/Model/CourseSyncSelectorInteractor.swift

```diff
@@ -0,0 +1,291 @@
+        courseSyncEntries
+            .map { $0.filter { $0.selectionState == .selected || $0.selectionState == .partiallySelected } }
+            .replaceError(with: [])
+            .first()
+            .eraseToAnyPublisher()
+    }
+
+    // MARK: - Private Methods
+
+    private func getTabs(courseId: String) -> AnyPublisher<[CourseSyncSelectorEntry.Tab], Error> {
+        ReactiveStore(
+            useCase: GetContextTabs(
+                context: Context.course(courseId)
+            )
+        )
+        .getEntities()
+        .map { $0.offlineSupportedTabs() }
+        .map {
+            $0.map {
+                CourseSyncSelectorEntry.Tab(
+                    id: "\(courseId)-\($0.id)",
+                    name: $0.label,
+                    type: $0.name
+                )
+            }
+        }
+        .eraseToAnyPublisher()
+    }
+
+    private func getAllFilesIfFilesTabIsEnabled(
+        course: CourseSyncSelectorCourse
+    ) -> AnyPublisher<CourseSyncSelectorEntry, Error> {
+        let tabs = Array(course.tabs).offlineSupportedTabs()
+        let mappedTabs = tabs.map {
+            CourseSyncSelectorEntry.Tab(
+                id: "\(course.courseId)-\($0.id)",
+                name: $0.label,
+                type: $0.name
+            )
+        }
+        if tabs.isFilesTabEnabled() {
+            return getAllFiles(courseId: course.courseId)
+                .map { files in
+                    CourseSyncSelectorEntry(
+                        name: course.name,
+                        id: course.courseId,
+                        tabs: mappedTabs,
+                        files: files
+                    )
+                }
+                .eraseToAnyPublisher()
+        } else {
+            return Just(
+                CourseSyncSelectorEntry(
+                    name: course.name,
+                    id: course.courseId,
+                    tabs: mappedTabs,
+                    files: []
+                )
+            )
+            .setFailureType(to: Error.self)
+            .eraseToAnyPublisher()
+        }
+    }
+
+    private func getAllFiles(courseId: String) -> AnyPublisher<[CourseSyncSelectorEntry.File], Error> {
+        unowned let unownedSelf = self
+
+        return ReactiveStore(
+            useCase: GetFolderByPath(
+                context: .course(courseId)
+            )
+        )
+        .getEntities()
+        .flatMap {
+            Publishers.Sequence(sequence: $0)
+                .filter { !$0.lockedForUser && !$0.hiddenForUser }
+                .setFailureType(to: Error.self)
+                .flatMap { unownedSelf.getFiles(folderID: $0.id, initialArray: []) }
+        }
+        .map {
+            $0.map {
+                CourseSyncSelectorEntry.File(
+                    id: $0.file?.id ?? Foundation.UUID().uuidString,
+                    name: $0.file?.displayName ?? "Unknown file",
+                    url: $0.file?.url
```

Comment:
I wonder if we could get an array of non optional files here. It seems that this `$0.file?` is on every line. Will it make sense to add a file with nil url?

Reply:
Fixed.